third_party/move/move-vm/types/src/loaded_data/runtime_types.rs

```diff
@@ -139,8 +140,16 @@ pub enum Type {
     Address,
     Signer,
     Vector(Box<Type>),
-    Struct(CachedStructIndex),
-    StructInstantiation(CachedStructIndex, Vec<Type>),
+    Struct {
+        index: CachedStructIndex,
+        ability: AbilitySet,
+    },
+    StructInstantiation {
+        index: CachedStructIndex,
+        ty_args: Vec<Type>,
+        base_ability: AbilitySet,
+        is_phantom_params: Vec<bool>,
+    },
```

Comment:
I'm a bit concerned by this change, since it reminds me of the `FatType` we painstakingly got rid of, but I also recognize that there aren't many other good ways to address the ability lookup problem you brought up, so I guess I'll be fine with this for now.

I feel the like the real problem with our current runtime type representation is that we try to make the type a self-contained struct. This means we either have to include a ton of information, potentially duplicate, in the struct itself (e.g. `FatType`) or have to perform expensive lookup (e.g. type to layout) if we want to keep the type itself lean.

A better representation might be that we store the actual type info in some kind of environment, and then expose some sort of super-lightweight references to the clients. Such references shall not only allow the client to access the information they need on-demand, but also make it possible for us to perform optimizations on the environment side, such as deduplication & lazy evaluation.

Reply:
I think there's a deeper cause here. The information of struct gets dispersed from the caller (struct handle) side and callee side (struct def) side. The full information can only be obtained by combining both which would require us to load the callee side as well.

Reply:
Could it be reasonable (as a future thing to consider) to disassociate this from the module storage, i.e. such that we could load type abilities without incurring the overhead of loading and processing everything else? Also, it does seem well-cacheable (smaller and not changing often?), and may benefit from not coupling with the whole corresponding modules and from sharing across blocks as well maybe in the future.